File: Game.cpp
#include "Game.h"
#include "Graphics.h"

bool dragvol=false;
bool dragsfx=false;
float musicvol=80;
float sfxvol=80;
bool muzik=false;

void Game::init() {
    srand(time(0));
    graphics.init();

    gMusic1 = graphics.loadMusic("assets/sound/retroarcadesound.mp3");
    gMusic=gMusic1;
    graphics.toggle(gMusic1);

    gMusic2=graphics.loadMusic("assets/sound/mario.mp3");
    gMusic3=graphics.loadMusic("assets/sound/bitgame.mp3");

    hup = graphics.loadChunk("assets/sound/hup-sound.mpga");
    meow = graphics.loadChunk("assets/sound/meow.mp3");
    pop = graphics.loadChunk("assets/sound/pop.mp3");

    background = graphics.loadTexture("assets/picture/background.jpg");
    playTexture = graphics.loadTexture("assets/picture/play.png");
    playagain = graphics.loadTexture("assets/picture/playagain.png");
    boardTexture = graphics.loadTexture("assets/picture/board.png");
    meodungyen = graphics.loadTexture("assets/picture/meodungyen.png");
    meodap = graphics.loadTexture("assets/picture/meodap.png");
    info = graphics.loadTexture("assets/picture/info.png");
    diemmeo = graphics.loadTexture("assets/picture/diemmeo.png");
    diemchuot = graphics.loadTexture("assets/picture/diemchuot.png");
    huongdan = graphics.loadTexture("assets/picture/huongdan.png");
    pausebutton=graphics.loadTexture("assets/picture/pause.png");
    menupause=graphics.loadTexture("assets/picture/manhinhpause.png");
    volumebar=graphics.loadTexture("assets/picture/volumebar.png");
    cursorvl=graphics.loadTexture("assets/picture/cursor.png");
    cursorsfx=cursorvl;
    sets=graphics.loadTexture("assets/picture/settings.png");
    menuset=graphics.loadTexture("assets/picture/manhinhset.png");
    tick=graphics.loadTexture("assets/picture/tick.png");

    for (int i = 0; i < 4; ++i) {
        anhsong[i] = graphics.loadTexture(("assets/picture/" + std::string(i == 0 ? "xamsong" : i == 1 ? "nausong" : i==2?"camsong":"xanh") + ".png").c_str());
        anhchet[i] = graphics.loadTexture(("assets/picture/" + std::string(i == 0 ? "xamchet" : i == 1 ? "nauchet" : i==2?"camchet":"xanh") + ".png").c_str());
    }

    xBasechuot[0] = 110; xBasechuot[1] = 170; xBasechuot[2] = 250; xBasechuot[3] = 320;
    xBasemeo[0] = 80;  xBasemeo[1] = 150; xBasemeo[2] = 220; xBasemeo[3] = 300;

    welcome = graphics.loadFont("assets/font/fluff.ttf", 100);
    gover = graphics.loadFont("assets/font/fluff.ttf", 80);
    diem = graphics.loadFont("assets/font/diem.ttf", 30);
    ingame = graphics.loadFont("assets/font/shortbaby.ttf", 20);
    opt=graphics.loadFont("assets/font/fluff.ttf",20);

    hello = graphics.renderText("Whack a", welcome, color);
    hellodongduoi = graphics.renderText("mouse!", welcome, color);
    gameover = graphics.renderText("Game over", gover, color);
    yourscore = graphics.renderText("Your score: ", diem, diemm);
    highscore = graphics.renderText("High score: ", diem, diemm);
    volume=graphics.renderText("Volume ",diem,diemm);
    sfx=graphics.renderText("Sound effect ",diem,diemm);
    mzopt=graphics.renderText("Music",diem,diemm);
    option=graphics.renderText("Retro         Mario        Bitgame",opt,{0,0,0,0});

    stat.loadHighScore();
}

void Game::updateSlider(){
    if(dragvol){
        float tmp=std::max(0,std::min(272,x-90));
        musicvol=tmp*128/272;
        Mix_VolumeMusic(musicvol);
    }
    int musicX=musicvol*272/128;
    graphics.renderTexture(cursorvl,std::min(342,std::max(90,90+musicX-10)),280);

    if(dragsfx){
        float tmp=std::max(0,std::min(272,x-90));
        sfxvol=tmp*128/272;
        Mix_Volume(-1,sfxvol);
    }
    int sfxX=sfxvol*272/128;
    graphics.renderTexture(cursorsfx,std::min(342,std::max(90,90+sfxX-10)),390);

}

void Game::updateMusic(){
    if(muzik){
        graphics.toggle(gMusic);
        if(m1.checkmouse(x,y)){
            graphics.renderTexture(tick,180,500);
            gMusic=gMusic1;
        }
        else if(m2.checkmouse(x,y)){
            graphics.renderTexture(tick,250,500);
            gMusic=gMusic2;
        }
        else if(m3.checkmouse(x,y)){
            graphics.renderTexture(tick,380,500);
            gMusic=gMusic3;
        }
        graphics.toggle(gMusic);
    }
    if(gMusic==gMusic1){
        graphics.renderTexture(tick,150,500);
    }
    if(gMusic==gMusic2){
        graphics.renderTexture(tick,250,500);
    }
    if(gMusic==gMusic3){
        graphics.renderTexture(tick,380,500);
    }
}

void Game::run() {
    init();
    bool isRunning = true;

    while (isRunning) {
        graphics.prepareScene(background);
        handleEvents();
        update();
        graphics.presentScene();
    }

    cleanUp();
}

void Game::handleEvents() {
    SDL_GetMouseState(&x, &y);
    while (SDL_PollEvent(&event)) {
        if (event.type == SDL_QUIT) {
            cleanUp();
            exit(0);
        }
        if (event.type == SDL_MOUSEBUTTONDOWN) {

            if (nutBatDau.checkmouse(x, y) && curState == State::StartScreen) {
                stat.reset();
                resetGame();
                curState = State::Playing;

            } else if(nutSet.checkmouse(x,y)&&curState==State::StartScreen){
                curState=State::Settings;
            }
            else if(curState==State::Settings){
                if(nutHome.checkmouse(x,y))
                    curState=State::StartScreen;
                else if(vlbar.checkmouse(x,y)) dragvol=true;
                else if(sfxbar.checkmouse(x,y)) dragsfx=true;
                else if(m1.checkmouse(x,y)||m2.checkmouse(x,y)||m3.checkmouse(x,y)){
                    muzik=true;
                }
            }

            else if (nutInfo.checkmouse(x, y) && curState == State::StartScreen) {
                curState = State::InfoScreen;
            } else if (tatInfo.checkmouse(x, y) && curState == State::InfoScreen) {
                curState = State::StartScreen;
            } else if (nutPlayAgain.checkmouse(x, y) && curState == State::GameOver) {
                stat.reset();
                resetGame();
                curState = State::Playing;

            } else if(curState==State::Playing)
            {
                for(int i=0;i<4;i++){
                    if(cot[i].checkmouse(x,y)){
                        cot[i].click=true;
                        cot[i].firsttime=true;
                        cot[i].clickTime = SDL_GetTicks();
                        cot[i].playingAnimation = true;
                        for(int j=0;j<4;j++){
                            if(i!=j){
                                cot[j].click=false;
                            }
                        }
                    }
                }
                if(nutPause.checkmouse(x,y)){
                    curState=State::Pause;
                }
            } else if(curState==State::Pause){
                if(nutHome.checkmouse(x,y))
                    curState=State::StartScreen;
                else if(nutResume.checkmouse(x,y)){
                    curState=State::Playing;
                }
                else if(vlbar.checkmouse(x,y)) dragvol=true;
                else if(sfxbar.checkmouse(x,y)) dragsfx=true;
            }
        }
        if(event.type==SDL_MOUSEBUTTONUP){
            dragvol=false;
            dragsfx=false;
            muzik=false;
        }
        if(event.type==SDL_KEYDOWN&&curState==State::Playing){
            if(event.key.keysym.sym==SDLK_ESCAPE){
                curState=State::Pause;
            }
        }
    }
}

void Game::resetGame() {
    mouse.clear();
    for (int i = 0; i < 4; ++i) {
        cot[i].click = false;
        cot[i].firsttime = true;
        cot[i].playingAnimation = false;
    }
    newGame=0;
    clicked = false;
    spawnInterval=3000;
}


void Game::update() {
    switch (curState) {
        case State::StartScreen:
            graphics.renderTexture(info, 10, 20);
            graphics.renderTexture(playTexture, 100, 550);
            graphics.renderTexture(hello, 20, 200);
            graphics.renderTexture(hellodongduoi, 80, 350);
            graphics.renderTexture(sets,400,10);
            break;

        case State::InfoScreen:
            graphics.renderTexture(huongdan, 60, 200);
            break;

        case State::Settings:
            graphics.renderTexture(menuset,60,200);
            graphics.renderTexture(volume,70,220);
            graphics.renderTexture(sfx,70,330);
            graphics.renderTexture(volumebar,90,280);
            graphics.renderTexture(volumebar,90,390);
            graphics.renderTexture(mzopt,70,440);
            graphics.renderTexture(option,70,500);
            updateMusic();
            updateSlider();
            break;

        case State::GameOver: {
            SDL_Texture* scoree = graphics.renderText(std::to_string(stat.score).c_str(), ingame, baby);
            SDL_Texture* mangg = graphics.renderText(std::to_string(stat.lives).c_str(), ingame, baby);
            SDL_Texture* hiscoree = graphics.renderText(std::to_string(stat.getHighScore()).c_str(), ingame, baby);

            graphics.renderTexture(playagain, 180, 550);
            graphics.renderTexture(gameover, 10, 200);
            graphics.renderTexture(yourscore, 100, 300);
            graphics.renderTexture(scoree, 300, 310);
            graphics.renderTexture(highscore, 100, 350);
            graphics.renderTexture(hiscoree, 300, 360);
            break;
        }

        case State::Playing: {
            if(stat.score>=30) spawnInterval=1500;
            else if(stat.score>=50) spawnInterval=1000;
            else if(stat.score>=80) spawnInterval=500;
            else if(stat.score>=100) spawnInterval=200;
            if(!newGame) {mouse.clear(); clicked=false;}
            graphics.renderTexture(diemmeo, 2, 50);
            graphics.renderTexture(diemchuot, 2, 10);
            graphics.renderTexture(graphics.renderText(std::to_string(stat.score).c_str(), ingame, baby), 60, 20);
            graphics.renderTexture(graphics.renderText(std::to_string(stat.lives).c_str(), ingame, baby), 60, 60);
            graphics.renderTexture(boardTexture, 0, 350);
            graphics.renderTexture(pausebutton,400,10);
            for (int i = 0; i < 4; i++) {
                if (cot[i].click) {
                    newGame++;
                    Uint32 currentTime = SDL_GetTicks();
                    if (cot[i].playingAnimation && currentTime - cot[i].clickTime <= 200) {
        // Hiện mèo đập trong 200ms
                        graphics.renderTexture(meodap, xBasemeo[i], 260);
                        if (cot[i].firsttime) {
                            Mix_PlayChannel(-1, hup, 0);
                            cot[i].firsttime = false;
                        }
                        for (auto& c : mouse) {
                            if (c.song && c.col == i && c.y <= 330&&c.y>=310) {
                                c.song=false;
                                if(c.speed==4) curState=State::GameOver;
                            }
                        }
                    } else {
        // Sau 200ms hiện mèo đứng yên
                        graphics.renderTexture(meodungyen, xBasemeo[i], 250);
                        cot[i].playingAnimation = false;
                    }
                    clicked=true;
                }
            }
            if(!clicked){
                graphics.renderTexture(meodungyen,190,250);
            }

            if (SDL_GetTicks() - lastSpawnTime >= spawnInterval) {
                int col = rand() % 4;
                int type;
                if(stat.score<30) type = rand() % 3;
                else type=rand()%4;
                Mouse c(col, 800, type + 1, anhsong[type], anhchet[type], true, true);
                mouse.push_back(c);
                lastSpawnTime = SDL_GetTicks();
            }



            for (auto& c : mouse) {
                if (c.visible) {
                    c.y -= c.speed;
                    if (c.y <= 300) {
                        c.visible = false;
                        if (c.song) {
                            if(c.speed!=4){
                                stat.loseLife();
                                Mix_PlayChannel(-1, meow, 0);
                            }
                        } else {
                            stat.addScore(c.speed);
                            Mix_PlayChannel(-1, pop, 0);
                        }
                    }
                }
            }


            for (auto& c : mouse) {
                if (!c.visible) continue;
                if (c.song)
                    graphics.renderTexture(c.texsong, xBasechuot[c.col], c.y);
                else
                    graphics.renderTexture(c.texchet, xBasechuot[c.col], c.y);
            }
            if(stat.lives<=0) {curState=State::GameOver;};
            break;
        }
        case State::Pause:
            graphics.renderTexture(menupause,60,200);
            graphics.renderTexture(volume,70,220);
            graphics.renderTexture(sfx,70,330);
            graphics.renderTexture(volumebar,90,280);
            graphics.renderTexture(volumebar,90,390);
            updateSlider();
            break;

    }
}

void Game::cleanUp() {
    stat.saveHighScore();
    graphics.quit();
}

File: web-version/src/Game.js
import { Graphics } from './Graphics.js';
import { Mouse } from './Mouse.js';
import { Column } from './Column.js';
import { Stat } from './Stat.js';
import { State } from './State.js';

export class Game {
    constructor(canvas) {
        this.graphics = new Graphics(canvas);
        this.stat = new Stat();
        this.curState = State.StartScreen;
        this.soundEnabled = true;
        this.newGame = true;
        this.clicked = false;
        this.spawnInterval = 2000;
        this.idx = 1;
        this.isRunning = true;
        this.lastSpawnTime = 0;
        this.musicX = 80;
        this.sfxX = 80;
        
        // Mouse positions
        this.xBasechuot = [110, 170, 250, 320];
        this.xBasemeo = [80, 150, 220, 300];
        
        // Game columns
        this.cot = [
            new Column(100, 170, 350, 800),
            new Column(175, 240, 350, 800),
            new Column(241, 313, 350, 800),
            new Column(314, 390, 350, 800)
        ];
        
        // UI buttons
        this.nutBatDau = new Column(100, 400, 550, 638);
        this.nutInfo = new Column(10, 60, 20, 70);
        this.tatInfo = new Column(360, 400, 200, 250);
        this.nutPlayAgain = new Column(100, 400, 550, 638);
        this.nutPause = new Column(400, 450, 10, 60);
        this.nutResume = new Column(150, 300, 480, 540);
        this.nutHome = new Column(150, 300, 570, 630);
        this.vlbar = new Column(90, 362, 280, 300);
        this.sfxbar = new Column(90, 362, 390, 410);
        this.nutSet = new Column(400, 450, 10, 55);
        this.m1 = new Column(150, 170, 500, 520);
        this.m2 = new Column(250, 270, 500, 520);
        this.m3 = new Column(380, 400, 500, 520);
        
        this.mouse = [];
        this.textures = {};
        this.sounds = {};
        this.music = {};
        
        this.dragvol = false;
        this.dragsfx = false;
        this.muzik = false;
    }

    async init() {
        // Load all textures
        this.textures.background = await this.graphics.loadTexture('assets/picture/background.jpg');
        this.textures.play = await this.graphics.loadTexture('assets/picture/play.png');
        this.textures.playagain = await this.graphics.loadTexture('assets/picture/playagain.png');
        this.textures.board = await this.graphics.loadTexture('assets/picture/board.png');
        this.textures.meodungyen = await this.graphics.loadTexture('assets/picture/meodungyen.png');
        this.textures.meodap = await this.graphics.loadTexture('assets/picture/meodap.png');
        this.textures.info = await this.graphics.loadTexture('assets/picture/info.png');
        this.textures.diemmeo = await this.graphics.loadTexture('assets/picture/diemmeo.png');
        this.textures.diemchuot = await this.graphics.loadTexture('assets/picture/diemchuot.png');
        this.textures.huongdan = await this.graphics.loadTexture('assets/picture/huongdan.png');
        this.textures.pause = await this.graphics.loadTexture('assets/picture/pause.png');
        this.textures.menupause = await this.graphics.loadTexture('assets/picture/manhinhpause.png');
        this.textures.volumebar = await this.graphics.loadTexture('assets/picture/volumebar.png');
        this.textures.cursor = await this.graphics.loadTexture('assets/picture/cursor.png');
        this.textures.settings = await this.graphics.loadTexture('assets/picture/settings.png');
        this.textures.menuset = await this.graphics.loadTexture('assets/picture/manhinhset.png');
        this.textures.tick = await this.graphics.loadTexture('assets/picture/tick.png');
        
        // Load mouse textures
        this.textures.anhsong = [];
        this.textures.anhchet = [];
        const mouseTypes = ['xamsong', 'nausong', 'camsong', 'xanh'];
        const mouseDead = ['xamchet', 'nauchet', 'camchet', 'xanh'];
        
        for (let i = 0; i < 4; i++) {
            this.textures.anhsong[i] = await this.graphics.loadTexture(`assets/picture/${mouseTypes[i]}.png`);
            this.textures.anhchet[i] = await this.graphics.loadTexture(`assets/picture/${mouseDead[i]}.png`);
        }
        
        // Load music
        this.music.retro = await this.graphics.loadMusic('assets/sound/retroarcadesound.mp3');
        this.music.mario = await this.graphics.loadMusic('assets/sound/mario.mp3');
        this.music.bitgame = await this.graphics.loadMusic('assets/sound/bitgame.mp3');
        this.currentMusic = this.music.retro;
        
        // Load sounds
        this.sounds.hup = await this.graphics.loadSound('assets/sound/hup-sound.mpga');
        this.sounds.meow = await this.graphics.loadSound('assets/sound/meow.mp3');
        this.sounds.pop = await this.graphics.loadSound('assets/sound/pop.mp3');
        
        this.graphics.playMusic(this.currentMusic);
    }

    handleEvents(event) {
        switch (event.type) {
            case 'click':
                this.handleClick(event.offsetX, event.offsetY);
                break;
            case 'keydown':
                if (event.key === 'Escape' && this.curState === State.Playing) {
                    this.curState = State.Pause;
                }
                break;
            case 'mousedown':
                this.handleMouseDown(event.offsetX, event.offsetY);
                break;
            case 'mouseup':
                this.handleMouseUp(event.offsetX, event.offsetY);
                break;
            case 'mousemove':
                this.handleMouseMove(event.offsetX, event.offsetY);
                break;
        }
    }

    handleClick(x, y) {
        switch (this.curState) {
            case State.StartScreen:
                if (this.nutBatDau.isInside(x, y)) {
                    this.curState = State.Playing;
                    this.resetGame();
                }
                if (this.nutInfo.isInside(x, y)) {
                    this.curState = State.InfoScreen;
                }
                if (this.nutSet.isInside(x, y)) {
                    this.curState = State.Settings;
                }
                break;
                
            case State.Playing:
                this.handleGameClick(x, y);
                if (this.nutPause.isInside(x, y)) {
                    this.curState = State.Pause;
                }
                break;
                
            case State.GameOver:
                if (this.nutPlayAgain.isInside(x, y)) {
                    this.curState = State.Playing;
                    this.resetGame();
                }
                break;
                
            case State.InfoScreen:
                if (this.tatInfo.isInside(x, y)) {
                    this.curState = State.StartScreen;
                }
                break;
                
            case State.Pause:
                if (this.nutResume.isInside(x, y)) {
                    this.curState = State.Playing;
                }
                if (this.nutHome.isInside(x, y)) {
                    this.curState = State.StartScreen;
                }
                break;
                
            case State.Settings:
                this.handleSettingsClick(x, y);
                break;
        }
    }

    handleGameClick(x, y) {
        for (let i = 0; i < this.mouse.length; i++) {
            if (this.cot[this.mouse[i].col].isInside(x, y) && this.mouse[i].visible && this.mouse[i].song) {
                this.mouse[i].song = false;
                this.mouse[i].texSong = this.textures.anhchet[this.mouse[i].col];
                
                // Different mouse types give different points
                if (this.mouse[i].col === 3) { // Green mouse - game over
                    this.stat.lives = 0;
                } else {
                    this.stat.score += (this.mouse[i].col + 1);
                    this.graphics.playSound(this.sounds.pop);
                }
                break;
            }
        }
    }

    handleSettingsClick(x, y) {
        if (this.m1.isInside(x, y)) {
            this.currentMusic = this.music.retro;
            this.graphics.playMusic(this.currentMusic);
        }
        if (this.m2.isInside(x, y)) {
            this.currentMusic = this.music.mario;
            this.graphics.playMusic(this.currentMusic);
        }
        if (this.m3.isInside(x, y)) {
            this.currentMusic = this.music.bitgame;
            this.graphics.playMusic(this.currentMusic);
        }
        if (this.nutHome.isInside(x, y)) {
            this.curState = State.StartScreen;
        }
    }

    handleMouseDown(x, y) {
        if (this.curState === State.Settings) {
            if (this.vlbar.isInside(x, y)) {
                this.dragvol = true;
            }
            if (this.sfxbar.isInside(x, y)) {
                this.dragsfx = true;
            }
        }
    }

    handleMouseUp(x, y) {
        this.dragvol = false;
        this.dragsfx = false;
        this.muzik = false;
    }

    handleMouseMove(x, y) {
        if (this.dragvol && this.curState === State.Settings) {
            const tmp = Math.max(0, Math.min(272, x - 90));
            this.musicX = tmp;
            this.graphics.setMusicVolume(tmp / 272);
        }
        if (this.dragsfx && this.curState === State.Settings) {
            const tmp = Math.max(0, Math.min(272, x - 90));
            this.sfxX = tmp;
            this.graphics.setSfxVolume(tmp / 272);
        }
    }

    update() {
        const currentTime = Date.now();
        
        if (this.curState === State.Playing) {
            // Spawn mice
            if (currentTime - this.lastSpawnTime > this.spawnInterval) {
                this.spawnMouseIfNeeded();
                this.lastSpawnTime = currentTime;
                
                // Increase difficulty
                if (this.spawnInterval > 800) {
                    this.spawnInterval -= 50;
                }
            }
            
            // Update mice
            for (let i = this.mouse.length - 1; i >= 0; i--) {
                const mouse = this.mouse[i];
                mouse.y += mouse.speed;
                
                // Remove mice that went off screen
                if (mouse.y > 800) {
                    if (mouse.song && mouse.col !== 3) { // Didn't catch mouse (except green)
                        this.stat.lives--;
                        this.graphics.playSound(this.sounds.meow);
                    }
                    this.mouse.splice(i, 1);
                }
            }
            
            // Check game over
            if (this.stat.lives <= 0) {
                this.curState = State.GameOver;
                this.stat.updateHighScore();
            }
        }
    }

    spawnMouseIfNeeded() {
        const col = Math.floor(Math.random() * 4);
        const mouseType = Math.floor(Math.random() * 100);
        let type, speed;
        
        if (mouseType < 50) {
            type = 0; speed = 2; // Gray mouse
        } else if (mouseType < 80) {
            type = 1; speed = 3; // Brown mouse  
        } else if (mouseType < 95) {
            type = 2; speed = 4; // Orange mouse
        } else {
            type = 3; speed = 4; // Green mouse (deadly)
        }
        
        const newMouse = new Mouse(
            col,
            350,
            speed,
            this.textures.anhsong[type],
            this.textures.anhchet[type],
            true,
            true
        );
        
        this.mouse.push(newMouse);
    }

    render() {
        this.graphics.clear();
        
        switch (this.curState) {
            case State.StartScreen:
                this.renderStartScreen();
                break;
            case State.Playing:
                this.renderGameScreen();
                break;
            case State.GameOver:
                this.renderGameOverScreen();
                break;
            case State.InfoScreen:
                this.renderInfoScreen();
                break;
            case State.Pause:
                this.renderPauseScreen();
                break;
            case State.Settings:
                this.renderSettingsScreen();
                break;
        }
    }

    renderStartScreen() {
        this.graphics.renderTexture(this.textures.background, 0, 0);
        this.graphics.renderTexture(this.textures.play, 100, 550);
        this.graphics.renderTexture(this.textures.info, 10, 20);
        this.graphics.renderTexture(this.textures.settings, 400, 10);
        this.graphics.renderText('WHACK-A-MOUSE', 150, 100, '40px Arial', '#fff');
        this.graphics.renderText(`High Score: ${this.stat.mxscore}`, 150, 150, '20px Arial', '#fff');
    }

    renderGameScreen() {
        this.graphics.renderTexture(this.textures.background, 0, 0);
        this.graphics.renderTexture(this.textures.board, 60, 350);
        this.graphics.renderTexture(this.textures.pause, 400, 10);
        
        // Render mice
        for (const mouse of this.mouse) {
            if (mouse.visible) {
                const texture = mouse.song ? mouse.texSong : mouse.texChet;
                const x = this.xBasechuot[mouse.col];
                this.graphics.renderTexture(texture, x, mouse.y, 50, 50);
            }
        }
        
        // Render UI
        this.graphics.renderText(`Score: ${this.stat.score}`, 20, 30, '20px Arial', '#fff');
        this.graphics.renderText(`Lives: ${this.stat.lives}`, 20, 60, '20px Arial', '#fff');
    }

    renderGameOverScreen() {
        this.graphics.renderTexture(this.textures.background, 0, 0);
        this.graphics.renderText('GAME OVER', 150, 200, '40px Arial', '#fff');
        this.graphics.renderText(`Final Score: ${this.stat.score}`, 150, 250, '24px Arial', '#fff');
        this.graphics.renderText(`High Score: ${this.stat.mxscore}`, 150, 280, '24px Arial', '#fff');
        this.graphics.renderTexture(this.textures.playagain, 100, 550);
    }

    renderInfoScreen() {
        this.graphics.renderTexture(this.textures.huongdan, 0, 0);
        this.graphics.renderTexture(this.textures.diemchuot, 360, 200);
        this.graphics.renderTexture(this.textures.diemmeo, 360, 200);
    }

    renderPauseScreen() {
        this.graphics.renderTexture(this.textures.menupause, 0, 0);
    }

    renderSettingsScreen() {
        this.graphics.renderTexture(this.textures.menuset, 0, 0);
        this.graphics.renderTexture(this.textures.volumebar, 90, 280);
        this.graphics.renderTexture(this.textures.volumebar, 90, 390);
        
        // Volume sliders
        const volX = Math.min(342, Math.max(90, 90 + this.musicX - 10));
        const sfxVolX = Math.min(342, Math.max(90, 90 + this.sfxX - 10));
        this.graphics.renderTexture(this.textures.cursor, volX, 280);
        this.graphics.renderTexture(this.textures.cursor, sfxVolX, 390);
        
        // Music options
        this.graphics.renderTexture(this.textures.tick, 150, 500);
    }

    resetGame() {
        this.stat.resetGame();
        this.mouse = [];
        this.spawnInterval = 2000;
        this.lastSpawnTime = Date.now();
    }

    run() {
        const gameLoop = () => {
            if (this.isRunning) {
                this.update();
                this.render();
                requestAnimationFrame(gameLoop);
            }
        };
        
        // Start game loop
        requestAnimationFrame(gameLoop);
        
        // Set up event listeners
        this.graphics.canvas.addEventListener('click', (e) => this.handleEvents({type: 'click', offsetX: e.offsetX, offsetY: e.offsetY}));
        this.graphics.canvas.addEventListener('mousedown', (e) => this.handleEvents({type: 'mousedown', offsetX: e.offsetX, offsetY: e.offsetY}));
        this.graphics.canvas.addEventListener('mouseup', (e) => this.handleEvents({type: 'mouseup', offsetX: e.offsetX, offsetY: e.offsetY}));
        this.graphics.canvas.addEventListener('mousemove', (e) => this.handleEvents({type: 'mousemove', offsetX: e.offsetX, offsetY: e.offsetY}));
        document.addEventListener('keydown', (e) => this.handleEvents({type: 'keydown', key: e.key}));
    }
}